src/Microsoft.FeatureManagement/FeatureFilters/TimeWindowFilter.cs

```diff
@@ -56,7 +69,83 @@ public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext context)
                 return Task.FromResult(false);
             }
 
-            return Task.FromResult((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value));
+            //
+            // Hit the first occurrence of the time window
+            if ((!settings.Start.HasValue || now >= settings.Start.Value) && (!settings.End.HasValue || now < settings.End.Value))
+            {
+                return Task.FromResult(true);
+            }
+
+            if (settings.Recurrence != null)
+            {
+                //
+                // The reference of the object will be used for hash key.
+                // If there is no pre-bounded settings attached to the context, there will be no cached filter settings and each call will have a unique settings object.
+                // In this case, the cache for recurrence settings won't work.
+                if (context.Settings == null)
+                {
+                    return Task.FromResult(RecurrenceEvaluator.MatchRecurrence(now, settings));
+                }
+
+                DateTimeOffset cachedTime = _recurrenceCache.GetOrAdd(
+                    settings,
+                    (_) =>
+                    {
+                        RecurrenceEvaluator.CalculateSurroundingOccurrences(now, settings, out DateTimeOffset? prevOccurrence, out DateTimeOffset? nextOccurrence);
+
+                        if (now < settings.Start.Value)
+                        {
+                            return nextOccurrence.Value;
+                        }
+
+                        if (prevOccurrence != null)
+                        {
+                            return prevOccurrence.Value;
+                        }
+
+                        //
+                        // There is no previous occurrence within the reccurrence range.
+                        return DateTimeOffset.MaxValue;
+                    });
+
+                if (now < cachedTime)
+                {
+                    return Task.FromResult(false);
+                }
+
+                if (now < cachedTime + (settings.End.Value - settings.Start.Value))
```

Comment:
Seems we should impose a limit on how long this can be. I'm not sure if we do that already.

Reply:
Now we will check whether the time window duration is shorter than how often it will recur. https://github.com/microsoft/FeatureManagement-Dotnet/pull/266#discussion_r1462221896

Reply:
How about add an additional check for the time window duration, e.g. the duration can not longer than 10 years?

Reply:
Do you think whether it is necessary to handle the potential System.ArgumentOutOfRangeException here? This operation may overflow the max datetime (i.e. 9999-12-31-23:59:59). But I don't believe this will happen in production.

Reply:
I think what we have now is fine then.